### ISSUE TITLE
Collision occurs the turn after

### ISSUE THREAD

**Issue by shummie** [HaliteChallenge/Halite-II#135]
See attached replay:
[replay-20171028-214016-0500--1898579245-312-208-90311.zip](https://github.com/HaliteChallenge/Halite-II/files/1424759/replay-20171028-214016-0500--1898579245-312-208-90311.zip)

On turn 32: ship id 11 is at position 128.79/90.74 and moves. On the web replay viewer, the ship is there. Now, when you advance 1 turn, the map object properties shows that the ship position for 11 is now at 133.79/91. However, if you look at the ship in the replay viewer, it doesn't actually move.

Advance 1 more frame, and then it looks like the ship crashes into Ship 9 which was docking onto the planet.

The distances between the two ships are 0.999974, so the ships should crash, but it's weird that 1) the replay viewer doesn't update the position of Ship 11 between frames 32 and 33, and then 2) the collision event occurs after the ship has already moved and is issuing a dock command when it should have probably crashed on frame 33 instead of 34.

Might be related to the below:
#114 
#129 

Note, this was tested locally, and i don't have the most up to date binaries as of 10/28/17. I can test with the same seed if a new binary is available to see if #129 fixes this.

**Comment by lidavidm** [comment]
Hmm, what I see in the replay is that the distance between ships 9 and 11 on frame 33 is 1.0000000000004252. The visualizer rounds the positions to be a little more friendly, but it can be misleading. The ships not moving is definitely a visualizer bug, though.

**Comment by lidavidm** [comment]
Ah wait, I just saw the discussion on Discord. I'm going to take a closer look once I find a good block of time…

**Comment by lidavidm** [comment]
@shummie , as far as I can tell, what's happening is:

- At the beginning of turn 33, ships 9 and 11 are alive, and so commands get issued. Ship 11 issues a dock command; ship 9 is already docking.
- They collide right away, and are destroyed on turn 33. (I'm basing this on the `events` field in the frame)
- The distance between them is slightly greater than 1 when they collide, and both should be stationary - looking into why that happens; it may be due to the residual velocity, if that got rounded down when writing the replay.

The visualizer doesn't appear to process the event until the end of the frame, for some reason. There is some wonkiness, especially when advancing frame by frame.

**Comment by shummie** [comment]
Yeah, your 3rd bullet point is the weird one. I downloaded the latest binary to see if the newest build which fixes some other things w/ velocity fixes it but I can still reproduce the same issue. Thanks for continuing to take a look at it! 

It's an ugly temporary hack but since velocity/inertia is not used this time, would an easy hack to just be set velocity to 0 at the end of each turn?

**Comment by lidavidm** [comment]
Yup, I can reproduce it as well, if I take over the replay you provided and issue the same commands starting at frame 29. I'm going to instrument the engine and take a peek at what's happening now. Thanks for the help reproducing this!

**Comment by lidavidm** [comment]
It turns out the replay is rounding things just enough to be misleading. Setting a breakpoint in GDB for when the ships collide:

```
(gdb) p id1
$9 = {_player_id = 0, _entity_index = 9, type = hlt::EntityType::ShipEntity}
(gdb) p id2
$10 = (const hlt::EntityId &) @0x95217c: {_player_id = 0, _entity_index = 11, 
  type = hlt::EntityType::ShipEntity}
(gdb) p ship1
$11 = (const hlt::Ship &) @0x955100: {<hlt::Entity> = {location = {pos_x = 134.78588000532, 
      pos_y = 91.052581378152098}, health = 255, radius = 0.5}, velocity = {vel_x = 0, 
    vel_y = 0}, weapon_cooldown = 0, docking_status = hlt::DockingStatus::Docking, 
  docking_progress = 1, docked_planet = 2}
(gdb) p ship2
$12 = (const hlt::Ship &) @0x954ee0: {<hlt::Entity> = {location = {
      pos_x = 133.78725047056543, pos_y = 91.000245421909185}, health = 255, radius = 0.5}, 
  velocity = {vel_x = 0, vel_y = 0}, weapon_cooldown = 0, 
  docking_status = hlt::DockingStatus::Docking, docking_progress = 5, docked_planet = 2}
```

The replay reports ship 11 is at `'x': 133.787250470565, 'y': 91.0002454219091` but the game sees `{ pos_x = 133.78725047056543, pos_y = 91.000245421909185}`, and similar for ship 9.

The distance between the two is 0.9999999999999951. This raises the question of why they don't collide at the end of the previous frame though.

**Comment by lidavidm** [comment]
Ah, in the previous turn, there is some slight rounding when ship 11's velocity is updated:

```
>>> decimal.Decimal('128.79410279679254') + decimal.Decimal('4.9931476737728691')
133.7872504705654091
>>> 128.79410279679254 + 4.9931476737728691
133.78725047056543
```

The ships are far apart enough in the previous frame that the engine decides they cannot intersect
```
(gdb) p distance
$23 = 6.000000000000008
(gdb) p ship2.velocity.magnitude()
$21 = 5
```
see for reference: https://github.com/HaliteChallenge/Halite-II/blob/94bfe5cdcf1d11378afca563a65e4c8d661e0d43/environment/core/SimulationEvent.cpp#L178

but after the position is updated, the rounding means that the ships end up colliding. Unrounded, they stay separated, as expected:
```
>>> y = decimal.Decimal('90.738565640694461') + decimal.Decimal('0.26167978121471913')
>>> x2 = decimal.Decimal('134.78588000532')
>>> y2 = decimal.Decimal('91.052581378152098')
>>> (x-x2)**2 + (y-y2)**2
1.000000000000031288195565158
```

**Comment by harikmenon** [comment]
@lidavidm is this something we need to fix?

**Comment by lidavidm** [comment]
@harikmenon If we could change the rounding modes to avoid things like this (as Janzert had suggested earlier), it would be nice to just avoid this edge case, though I'd worry that edge cases like this will keep cropping up.

**Comment by harikmenon** [comment]
Do we know what the impact is going to be?

**Comment by lidavidm** [comment]
At least here, it just means that the ship is killed on the start of the turn after it dies instead of the end of the turn that it should actually die. That might mean it could potentially attack one extra time.

**Comment by harikmenon** [comment]
I think we should fix this then.

**Comment by j-clap** [comment]
This shouldn't be an issue anymore since we have full precision now.

**Comment by lidavidm** [comment]
This is actually unrelated: this is referring to a calculation within the engine itself, which is rounded incorrectly.

**Comment by harikmenon** [comment]
@lidavidm what is the consensus here, are we fixing this?

**Comment by lidavidm** [comment]
@harikmenon This case at least should have been fixed in #319.

**Comment by harikmenon** [comment]
Sweet. Closing this then.
